store/tikv/gcworker/gc_worker.go

```diff
@@ -422,6 +423,127 @@ func (w *GCWorker) deleteRanges(ctx goctx.Context, safePoint uint64) error {
 	return nil
 }
 
+type gcTask struct {
+	startKey  []byte
+	endKey    []byte
+	safePoint uint64
+}
+
+type gcResult struct {
+	task           *gcTask
+	successRegions int
+	failedRegions  int
+	err            error
+}
+
+func getNextGCTask(store tikv.Storage, bo *tikv.Backoffer, safePoint uint64, lastKey kv.Key) (*gcTask, error) {
+	loc, err := store.GetRegionCache().LocateKey(bo, lastKey)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	task := &gcTask{
+		startKey:  lastKey,
+		endKey:    loc.EndKey,
+		safePoint: safePoint,
+	}
+	return task, nil
+}
+
+func doGC(ctx goctx.Context, store tikv.Storage, safePoint uint64, identifier string) error {
+	gcWorkerCounter.WithLabelValues("do_gc").Inc()
+
+	err := saveSafePoint(store.GetSafePointKV(), tikv.GcSavedSafePoint, safePoint)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Sleep to wait for all other tidb instances update their safepoint cache.
+	time.Sleep(gcSafePointCacheInterval)
+
+	successRegions := 0
+	failedRegions := 0
+	remained := 0
+	startTime := time.Now()
+	log.Infof("[gc worker] %s start gc, safePoint: %v.", identifier, safePoint)
+	defer func() {
+		log.Infof("[gc worker] %s finish gc, running jobs %v, safePoint: %v, successful regions: %v,"+
+			" failed regions: %v, cost time: %s",
+			identifier, remained, safePoint, successRegions, failedRegions, time.Since(startTime))
+		gcHistogram.WithLabelValues("do_gc").Observe(time.Since(startTime).Seconds())
+	}()
+
+	var key []byte
+	gcResultChan := make(chan *gcResult, gcDefaultGCConcurrency)
```

Comment:
gcTaskQueue := make(chan gcTask, gcDefaultGCConcurrency)
gcTaskCh := make(chan gcTask, 1024)
for i = 0; i < gcDefaultGCConcurrency; i++ {
 go runGcTaskWorker(gcTaskQueue, gcResultCh)
}
var key []byte
for {
  task, err := genNextGCTask(...)
  if err != nil {...}
  if task == nil { 
     send N nil to gcTaskQueue, notify all workers to quit
     break;
  }
  gcTaskQueue <- task

  ...receive task results if have
}
reveive task results util N nil